backend/danswer/connectors/gmail/connector.py

```diff
@@ -0,0 +1,234 @@
+            )
+            creds = get_gmail_creds_for_authorized_user(
+                token_json_str=access_token_json_str
+            )
+
+            # tell caller to update token stored in DB if it has changed
+            # (e.g. the token has been refreshed)
+            new_creds_json_str = creds.to_json() if creds else ""
+            if new_creds_json_str != access_token_json_str:
+                new_creds_dict = {DB_CREDENTIALS_DICT_TOKEN_KEY: new_creds_json_str}
+
+        if GMAIL_DB_CREDENTIALS_DICT_SERVICE_ACCOUNT_KEY in credentials:
+            service_account_key_json_str = credentials[
+                GMAIL_DB_CREDENTIALS_DICT_SERVICE_ACCOUNT_KEY
+            ]
+            creds = get_gmail_creds_for_service_account(
+                service_account_key_json_str=service_account_key_json_str
+            )
+
+            # "Impersonate" a user if one is specified
+            delegated_user_email = cast(
+                str | None, credentials.get(DB_CREDENTIALS_DICT_DELEGATED_USER_KEY)
+            )
+            if delegated_user_email:
+                creds = creds.with_subject(delegated_user_email) if creds else None
+
+        if creds is None:
+            raise PermissionError(
+                "Unable to access Gmail - unknown credential structure."
+            )
+
+        self.creds = creds
+        return new_creds_dict
+
+    @staticmethod
+    def _strptime_to_datetime(_input: str) -> datetime:
+        _input = _input.split(", ")[-1]  # remove weekday if exist
+        _input = _input.rsplit(" ", 1)[0]  # remove timezone: GMT, (GMT), (UTC), etc.
+        if "+" not in _input and "-" not in _input:
+            return datetime.strptime(_input, "%d %b %Y %H:%M:%S")
+        return datetime.strptime(_input, "%d %b %Y %H:%M:%S %z")
+
+    def _get_email_body(self, payload) -> str:
+        parts = payload.get("parts", [])
+        email_body = ""
+        for part in parts:
+            mime_type = part.get("mimeType")
+            body = part.get("body")
+            if mime_type == "text/plain":
+                data = body.get("data", "")
+                text = urlsafe_b64decode(data).decode()
+                email_body += text
+        return email_body
+
+    def _email_to_document(self, full_email: Dict[str, Any]) -> Document:
+        email_id = full_email["id"]
+        payload = full_email["payload"]
+        headers = payload.get("headers")
+        if headers:
+            for header in headers:
+                name = header.get("name")
+                value = header.get("value")
+                if name.lower() == "from":
+                    _from = value
+                if name.lower() == "to":
+                    _to = value
+                if name.lower() == "subject":
+                    _subject = value
+                if name.lower() == "date":
+                    _date = value
+        email_data = f"{_from} {_to} {_subject} {_date}"
```

Comment:
Might help the LLM to have some text prefix like "From:", "To:" etc, unless that's already included?

Reply:
True that! added - thanks :)